src/rust/client-executor/src/lib.rs

```diff
@@ -0,0 +1,421 @@
+    I: Iterator<Item = Duration>,
+    A: Action,
+{
+    strategy: I,
+    #[pin]
+    state: ExecuteState<A>,
+    recloser: &'a AsyncRecloser,
+    timeout: Duration,
+    action: A,
+}
+
+impl<'a, I, A> Execution<'a, I, A>
+where
+    I: Iterator<Item = Duration>,
+    A: Action,
+{
+    fn attempt(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<ActionResult<A>> {
+        let future = {
+            let this = self.as_mut().project();
+
+            this.recloser
+                .call(TryTimeout::new(this.timeout.clone(), this.action.run()))
+        };
+        self.as_mut()
+            .project()
+            .state
+            .set(ExecuteState::Running(future));
+        self.poll(cx)
+    }
+
+    fn retry(
+        mut self: Pin<&mut Self>,
+        err: Error<A::Error>,
+        cx: &mut Context,
+    ) -> Result<Poll<ActionResult<A>>, Error<<A as Action>::Error>> {
+        match self.as_mut().project().strategy.next() {
+            None => {
+                tracing::debug!("No more retries");
+                Err(err)
+            }
+            Some(duration) => {
+                tracing::debug!("Retrying in {}ms", duration.as_millis());
+                let deadline = Instant::now() + duration;
+                let future = sleep_until(deadline);
+                self.as_mut()
+                    .project()
+                    .state
+                    .set(ExecuteState::Sleeping(future));
+                Ok(self.poll(cx))
+            }
+        }
+    }
+}
+
+impl<'a, I, A> Future for Execution<'a, I, A>
+where
+    I: Iterator<Item = Duration>,
+    A: Action,
+{
+    type Output = ActionResult<A>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        match self.as_mut().project().state.poll(cx) {
+            ExecuteFuturePoll::Running(poll_result) => match poll_result {
+                Poll::Ready(Ok(ok)) => Poll::Ready(Ok(ok)),
+                Poll::Pending => Poll::Pending,
+                Poll::Ready(Err(err)) => match self.retry(err, cx) {
+                    Ok(poll) => poll,
+                    Err(err) => Poll::Ready(Err(err)),
+                },
+            },
+            ExecuteFuturePoll::Sleeping(poll_result) => match poll_result {
+                Poll::Pending => Poll::Pending,
+                Poll::Ready(_) => self.attempt(cx),
+            },
+        }
+    }
+}
+
+/// Based on https://resilience4j.readme.io/docs/circuitbreaker
+pub struct ExecutorConfig {
+    builder: RecloserBuilder,
+    timeout: Duration,
+}
+
+impl ExecutorConfig {
+    /// Creates a new `ExecutorConfig` from a `timeout`. Note that the `timeout`
+    /// is *not* a global timeout, but will be applied to each individual call
+    /// of the underlying future.
+    /// The rest of the parameters - threshold, closed_len, etc, are initialized
+    /// to defaults.
+    pub fn new(timeout: Duration) -> Self {
+        let builder = Recloser::custom()
+            .error_rate(0.5)
```

Comment:
Why 0.5? Similar questions apply to the other hardcoded numbers here.

Reply:
https://resilience4j.readme.io/docs/circuitbreaker

They're just based on the resilience4j defaults, or very slightly modified (I decreased the CircuitOpen timeout).